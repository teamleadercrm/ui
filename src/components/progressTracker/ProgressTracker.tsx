import React, { ReactNode } from 'react';
import cx from 'classnames';
import theme from './theme.css';

import Box from '../box';
import ProgressStep from './ProgressStep';
import { GenericComponent } from '../../@types/types';
import { COLORS } from '../../constants';

export interface ProgressTrackerProps {
  /** Whether or not all steps are completed */
  done?: boolean;
  /** The number of the step which is currently active */
  currentStep?: number;
  /** The steps to display inside the progress tracker */
  children?: ReactNode;
  /** Color theme of the progress tracker. */
  color?: (typeof COLORS)[number];
  /** Where to position the labels. Alternating allows for wider labels. */
  labelPosition?: 'top' | 'alternating' | 'bottom';
}

const ProgressTracker: GenericComponent<ProgressTrackerProps> & { ProgressStep: typeof ProgressStep } = ({
  color = 'mint',
  children,
  currentStep = 0,
  done,
  labelPosition = 'top',
}: ProgressTrackerProps) => {
  const classNames = cx(theme['tracker'], theme[color], theme[`tracker-${labelPosition}`]);

  return (
    <Box data-teamleader-ui="progress-tracker" className={classNames}>
      {React.Children.map(children, (child, index) => {
        const activeStep = Math.max(0, currentStep);
        const childProps = React.isValidElement(child) && child.props;

        return (
          <ProgressStep
            active={done ? false : index === activeStep}
            completed={done || index < activeStep}
            {...childProps}
          />
        );
      })}
    </Box>
  );
};

ProgressTracker.ProgressStep = ProgressStep;

export default ProgressTracker;
