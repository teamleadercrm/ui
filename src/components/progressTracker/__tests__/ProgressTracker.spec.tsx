import React from 'react';
import { render } from '@testing-library/react';
import ProgressTracker from '../ProgressTracker';

describe('Component - ProgressTracker', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <ProgressTracker>
        <ProgressTracker.ProgressStep label="New" />
        <ProgressTracker.ProgressStep label="In progress" />
        <ProgressTracker.ProgressStep label="Done" />
      </ProgressTracker>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
