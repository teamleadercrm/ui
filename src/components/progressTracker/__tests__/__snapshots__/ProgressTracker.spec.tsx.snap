// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - ProgressTracker renders 1`] = `
<DocumentFragment>
  <div
    class="box tracker mint tracker-top"
    data-teamleader-ui="progress-tracker"
  >
    <div
      class="box step is-active"
    >
      <div
        class="step-label-holder"
      >
        <p
          class="box text text-small darkest step-label"
          data-teamleader-ui="text"
        >
          New
        </p>
      </div>
      <span
        class="status-bullet-halo"
      />
      <span
        class="status-bullet"
      />
    </div>
    <div
      class="box step"
    >
      <div
        class="step-label-holder"
      >
        <p
          class="box text text-small darkest step-label"
          data-teamleader-ui="text"
        >
          In progress
        </p>
      </div>
      <span
        class="status-bullet-halo"
      />
      <span
        class="status-bullet"
      />
    </div>
    <div
      class="box step"
    >
      <div
        class="step-label-holder"
      >
        <p
          class="box text text-small darkest step-label"
          data-teamleader-ui="text"
        >
          Done
        </p>
      </div>
      <span
        class="status-bullet-halo"
      />
      <span
        class="status-bullet"
      />
    </div>
  </div>
</DocumentFragment>
`;
